x/superfluid/keeper/stake.go

```diff
@@ -656,3 +606,217 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	lock, err := k.lk.GetLockByID(ctx, lockId)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	// check lock owner is sender
+	if lock.Owner != sender.String() {
+		return sdk.ZeroInt(), types.LockOwnerMismatchError{
+			LockId:        lock.ID,
+			LockOwner:     lock.Owner,
+			ProvidedOwner: sender.String(),
+		}
+	}
+
+	lockCoin := lock.Coins[0]
+
+	// Ensuring the sharesToMigrate contains gamm pool share prefix.
+	if !strings.HasPrefix(lockCoin.Denom, gammtypes.GAMMTokenPrefix) {
+		return sdk.Int{}, types.SharesToMigrateDenomPrefixError{Denom: lockCoin.Denom, ExpectedDenomPrefix: gammtypes.GAMMTokenPrefix}
+	}
+
+	poolIdLeaving, err := gammtypes.GetPoolIdFromShareDenom(lockCoin.Denom)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	var superfluidValAddr string
+	interAcc, found := k.GetIntermediaryAccountFromLockId(ctx, lockId)
+	if found {
+		superfluidValAddr = interAcc.ValAddr
+	}
+
+	// Force unlock, validate the provided sharesToStake, and exit the balancer pool.
+	// we exit with min token out amount zero since we are checking min amount designated to stake later on anyways.
+	exitCoins, err := k.forceUnlockAndExitBalancerPool(ctx, sender, poolIdLeaving, lock, lockCoin, sdk.NewCoins(), false)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	totalAmtConverted, err = k.convertGammSharesToOsmoAndStake(ctx, sender, valAddr, poolIdLeaving, exitCoins, minAmtToStake, superfluidValAddr)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	return totalAmtConverted, nil
+}
+
+func (k Keeper) convertUnlockedToStake(ctx sdk.Context, sender sdk.AccAddress, valAddr string, sharesToStake sdk.Coin,
+	minAmtToStake sdk.Int) (totalAmtConverted sdk.Int, err error) {
+	if !strings.HasPrefix(sharesToStake.Denom, gammtypes.GAMMTokenPrefix) {
+		return sdk.Int{}, types.SharesToMigrateDenomPrefixError{Denom: sharesToStake.Denom, ExpectedDenomPrefix: gammtypes.GAMMTokenPrefix}
+	}
+
+	// Get the balancer poolId by parsing the gamm share denom.
+	poolIdLeaving, err := gammtypes.GetPoolIdFromShareDenom(sharesToStake.Denom)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	// Exit the balancer pool position.
+	// we exit with min token out amount zero since we are checking min amount designated to stake later on anyways.
+	exitCoins, err := k.gk.ExitPool(ctx, sender, poolIdLeaving, sharesToStake.Amount, sdk.NewCoins())
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	totalAmtConverted, err = k.convertGammSharesToOsmoAndStake(ctx, sender, valAddr, poolIdLeaving, exitCoins, minAmtToStake, "")
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	return totalAmtConverted, nil
+}
+
+// convertGammSharesToOsmoAndStake converts given gamm shares to osmo by swapping in the given pool
+// then stakes it to the designated validator.
+// minAmtToStake works as slippage bound, and would error if total amount being staked is less than min amount to stake.
+// Acutal delegation is done based on input paramters of delegateBaseOnValsetPref.
```

Comment:
Please also elaborate on parameter values. For example, `originalSuperfluidValAddr` can be an empty string which is unintuitive

Reply:
Added in https://github.com/osmosis-labs/osmosis/pull/5949/commits/18f4bd7e81d8dd238fc541fa4d9cd840ca82d0ed